thorlcr/master/thmastermain.cpp

```diff
@@ -338,27 +338,86 @@ class CRegistryServer : public CSimpleInterface
     {
         CriticalBlock block(crit);
         unsigned i=0;
-        for (; i<queryNodeClusterWidth(); i++)
+        unsigned numNodes = queryNodeClusterWidth();
+        unsigned numReplyNodes = 0;
+        mptag_t shutdownTag = createReplyTag();
+        for (; i<numNodes; i++)
         {
             if (status->test(i))
             {
-                status->set(i, false);
                 SocketEndpoint ep = queryNodeGroup().queryNode(i+1).endpoint();
-                if (watchdog)
-                    watchdog->removeSlave(ep);
                 CMessageBuffer msg;
                 msg.append((unsigned)Shutdown);
+                msg.append((unsigned)shutdownTag);
+#ifdef _DEBUG
+                StringBuffer tmp;
+                ep.getUrlStr(tmp);
+                DBGLOG("sending Shutdown:%u msg to slave %d (%s)", (unsigned)shutdownTag, i+1, tmp.str());
+#endif
                 try
                 {
-                    queryNodeComm().send(msg, i+1, masterSlaveMpTag, MP_ASYNC_SEND);
+                    if (queryNodeComm().send(msg, i+1, masterSlaveMpTag, MP_ASYNC_SEND))
+                        numReplyNodes++;
+                    else
+                        status->set(i, false);
+                }
+                catch (IMP_Exception *e)
+                {
+                    e->Release();
+                    status->set(i, false);
                 }
-                catch (IMP_Exception *e) { e->Release(); }
                 catch (IException *e)
                 {
-                    EXCLOG(e, "Shutting down slave");
+                    EXCLOG(e, "Shutting down slave - send");
                     e->Release();
+                    status->set(i, false);
+                }
+                if (watchdog)
+                    watchdog->removeSlave(ep);
+            }
+        }
+
+        CTimeMon tm(20000);
+        unsigned numReplied = 0;
+        while (numReplied < numReplyNodes)
+        {
+            unsigned remaining;
+            if (tm.timedout(&remaining))
+            {
+                PROGLOG("Timeout waiting for Shutdown reply from slave(s) (%u replied out of %u total)", numReplied, numReplyNodes);
+                break;
+            }
+            i = 0;
+            for (; i<numNodes; i++)
+            {
+                if (status->test(i))
+                {
+                    try
+                    {
+                        rank_t sender;
+                        CMessageBuffer msg;
+                        if (queryNodeComm().recv(msg, i+1, shutdownTag, &sender, 100))
```

Comment:
It would be better to not force receipt order here.. and listen to RANK_ALL
give a warning if from an unexpected source.

So outer loop here would be for count of slavesRegistered.